packages/itwinui-react/src/core/utils/functions/dev.ts

```diff
@@ -2,14 +2,21 @@
  * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-const isJest = typeof (globalThis as any).jest !== 'undefined';
+const isTestingFramework = () => {
+  return (
+    typeof vitest !== 'undefined' ||
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    typeof (globalThis as any).jest !== `undefined` ||
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    typeof (globalThis as any).mocha !== 'undefined'
```

Comment:
Have you actually tested or researched that this is the correct way of detecting mocha?

We can't just check `globalThis.whatever`. We know from experience that it works for `jest` and `vitest` because both of them inject globals, but `mocha` might not do the same.

Reply:
Can the check for mocha come later when we're using Mocha?

Reply:
In that case, instead of testing whether we're in a testing framework, can we test if we're in the browser? If so, that will be more reliable since users may use non-jest/vitest/mocha testing frameworks also.

Reply:
Detecting the browser serves a different purpose, and our components can be SSR'd (using one of [React's server apis](https://react.dev/reference/react-dom/server)).

At best we could try to detect if the code is running in `jsdom`, but then that excludes alternative DOM implementations like `happy-dom`.

Plus, both vitest and mocha can run in the browser too. We might want to still bail if it's running in the browser?

Reply:
Upon thinking about it, I just realized, I think our goal should be to highlight *all* incorrect uses of our components. The only exception is when in production.

So, if a user's test case uses our components incorrectly, should we show the console warning when running the unit tests too? I feel that is the safest, and also easier on our end.

```js
// Show dev errors in all developer-side environments. i.e. dev server/browser, testing environments, etc.
isDev = process.env.NODE_ENV !== 'production';
```

Reply:
> Upon thinking about it, I just realized, I think our goal should be to highlight all incorrect uses of our components. The only exception is when in production.

It's not just about highlighting incorrect usages. We straight up don't want to run certain code in test environments. See current usage in `ThemeProvider`.

Also I think it would be kinda annoying to surface dev-only warnings during tests. The terminal could get flooded with lots of warnings. Although, maybe that's fine? It would encourage users to fix those warnings. Regardless, we need separate checks for reasons mentioned above.

> Both the assert statement returns false in both cases when jest/mocha is installed or not. Am I testing something wrong? Need suggestions.

I don't understand what you're testing. The task here is to research what is the correct way to detect mocha, and then test that it works by running that code inside a mocha test. The jest check is already fine and doesn't need to be touched.

Reply:
From my perfunctory research, this is the best I've found: https://github.com/JoshuaKGoldberg/console-fail-test/blob/main/src/environments/mocha.ts#L19-L28

via https://redirect.github.com/mochajs/mocha/issues/5084

Maybe it would be enough to check `typeof beforeEach !== "undefined"` or `typeof describe !== "undefined"`, provided it doesn't trigger for Cypress or Playwright.

Reply:
Lets go with this for now. Can be improved later.

```js
const isMocha =
  typeof globalThis.beforeEach !== "undefined" &&
    `${globalThis.beforeEach}`.replace(/\s/g, "") ===
      "function(name,fn){suites[0].beforeEach(name,fn);}"
```

Reply:
I just followed their [getting started guide](https://mochajs.org/#getting-started) in a new project: https://stackblitz.com/edit/stackblitz-starters-93pyor?file=test%2Ftest.test.js